pdns/dnsdistdist/doq.cc

```diff
@@ -0,0 +1,875 @@
+                                                   (struct sockaddr*) &peer,
+                                                   peer.getSocklen(),
+                                                   config.get()), quiche_conn_free);
+  auto conn = Connection(peer, std::move(quicheConn));
+  auto pair = s_connections.emplace(serverSideID,  std::move(conn));
+  return pair.first->second;
+}
+
+static void flushEgress(Socket& sock, Connection& conn) {
+  std::array<uint8_t, MAX_DATAGRAM_SIZE> out;
+  quiche_send_info send_info;
+
+  while (true) {
+    auto written = quiche_conn_send(conn.d_conn.get(), out.data(), out.size(), &send_info);
+
+    if (written == QUICHE_ERR_DONE) {
+      return;
+    }
+
+    if (written < 0) {
+      return;
+    }
+
+    sock.sendTo(reinterpret_cast<const char*>(out.data()), written, conn.d_peer);
+  }
+
+  // FIXME: update timers
+  // -> on peut appeler une fonction quiche pour savoir quand prochain timeout
+  // -> pas ici ?
+  // -> fin de loop event quand est le prochain plus petit timeout a venir
+  // -> relancer le multiplexer pour au plus ce temps la
+}
+
+std::unique_ptr<CrossProtocolQuery> getDOQCrossProtocolQueryFromDQ(DNSQuestion& dq, bool isResponse)
+{
+  if (!dq.ids.doqu) {
+    throw std::runtime_error("Trying to create a DoQ cross protocol query without a valid DoQ unit");
+  }
+
+  auto du = std::move(dq.ids.doqu);
+  if (&dq.ids != &du->ids) {
+   du->ids = std::move(dq.ids);
+  }
+
+  du->ids.origID = dq.getHeader()->id;
+
+  if (!isResponse) {
+    if (du->query.data() != dq.getMutableData().data()) {
+      du->query = std::move(dq.getMutableData());
+    }
+  }
+  else {
+    if (du->response.data() != dq.getMutableData().data()) {
+      du->response = std::move(dq.getMutableData());
+    }
+  }
+
+  return std::make_unique<DOQCrossProtocolQuery>(std::move(du), isResponse);
+}
+
+/*
+   We are not in the main DoQ thread but in the DoQ 'client' thread.
+*/
+static void processDOQQuery(DOQUnitUniquePtr&& unit, bool inMainThread = false)
+{
+  const auto handleImmediateResponse = [inMainThread](DOQUnitUniquePtr&& du, const char* reason) {
+    DEBUGLOG("handleImmediateResponse() reason=" << reason);
+    if (inMainThread) {
+      auto conn = getConnection(du->serverConnID);
+      handleResponse(*du->dsc->df, *conn, du->streamID, du->response);
+      /* so the unique pointer is stored in the InternalState which itself is stored in the unique pointer itself. We likely need
+         a better design, but for now let's just reset the internal one since we know it is no longer needed. */
+      du->ids.doqu.reset();
+    }
+    else {
+      sendDOQUnitToTheMainThread(std::move(du), reason);
+    }
+  };
+
+  auto& ids = unit->ids;
+  ids.doqu = std::move(unit);
+  auto& du = ids.doqu;
+  uint16_t queryId = 0;
+  ComboAddress remote;
+
+  try {
+    {
+      // if there was no EDNS, we add it with a large buffer size
+      // so we can use UDP to talk to the backend.
```

Comment:
I'm not sure it is a good idea to do this over QUIC, it depends on how/if QUIC handles packets larger than the MTU.